packages/cel/src/equals.ts

```diff
@@ -0,0 +1,222 @@
+// Copyright 2024-2025 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { isMessage } from "@bufbuild/protobuf";
+import {
+  CelList,
+  CelMap,
+  CelObject,
+  CelType,
+  CelUint,
+  ProtoNull,
+} from "./value/value.js";
+import {
+  isReflectList,
+  isReflectMap,
+  isReflectMessage,
+  reflect,
+  type ReflectList,
+  type ReflectMap,
+} from "@bufbuild/protobuf/reflect";
+import { getEvalContext } from "./eval.js";
+import { equals as equalsMessage } from "@bufbuild/protobuf";
+import {
+  DurationSchema,
+  isWrapper,
+  TimestampSchema,
+} from "@bufbuild/protobuf/wkt";
+
+/**
+ * Checks for equality of two CEL values. It follows the following rules:
+ *
+ * - Primitives are checked for value equality using native equality check.
+ * - Types need not be equal for numerical values to be equal.
+ * - Wrappers are treated as scalars.
+ * - Protobuf messages are checked using protobuf equality
+ * - Lists are equal of if lengths match and each element matches the corresponding element in the other one.
+ * - Maps are equal if both the have the same set of keys and corresponding values.
+ *
+ * If the two cannot be compared, it returns false.
+ */
+export function equals(lhs: unknown, rhs: unknown): boolean {
+  // Fast path for Int, Double, Bool, and String types.
+  if (lhs === rhs) {
+    return true;
+  }
+  // Remaining scalars or scalars from Wrapper types
+  if (lhs instanceof CelUint || (isMessage(lhs) && isWrapper(lhs))) {
+    lhs = lhs.value;
+  }
+  if (rhs instanceof CelUint || (isMessage(rhs) && isWrapper(rhs))) {
+    rhs = rhs.value;
+  }
+  // Do a simple check again for uint/wrapper types.
+  if (lhs === rhs) {
+    return true;
+  }
+  // Check for scalars that are of different type and unwrapped values
+  if (
+    (typeof lhs === "number" || typeof lhs === "bigint") &&
+    (typeof rhs === "number" || typeof rhs === "bigint")
+  ) {
+    if (typeof lhs !== typeof rhs) {
+      lhs = Number(lhs);
+      rhs = Number(rhs);
+    }
```

Comment:
I don't think we should coerce to the more narrow type. The comparison operator `==` is a bad idea for comparing a string to a bool, but it has the behavior we need when applied to number and/or bigint:

```suggestion
    return lhs == rhs;
```